File: cloudflare-worker/src/predictors/strategies/qxc/frequency.js
/**
 * 七星彩频率策略
 */

import { BaseStrategy } from './base.js';

export class FrequencyStrategy extends BaseStrategy {
  constructor() {
    super('frequency', '频率策略', '基于历史出现频率，选择高频号码');
  }

  generateNumbers(context) {
    const { numberFrequency } = context;
    const numbers = [];
    
    if (numberFrequency && numberFrequency.length > 0) {
      // 按位置选择高频号码（带随机性）
      for (let i = 0; i < 7; i++) {
        // 从前5个高频号码中随机选择
        const topNumbers = numberFrequency.slice(0, 5);
        const selected = this.randomChoice(topNumbers);
        numbers.push(parseInt(selected.ball));
      }
    } else {
      // 如果没有频率数据，随机生成
      for (let i = 0; i < 7; i++) {
        numbers.push(this.randomNumber());
      }
    }
    
    return numbers;
  }
}

File: lotteries/qxc/strategies/frequency.py
"""
频率策略 - 基于每个位置的历史出现频率
"""

from .base import BaseStrategy
from typing import List, Dict
import random


class FrequencyStrategy(BaseStrategy):
    """频率策略：基于每个位置的历史频率选择号码"""
    
    def __init__(self):
        super().__init__(
            name='频率策略',
            description='基于每个位置的历史出现频率选择号码'
        )
    
    def generate_numbers(self, context: Dict) -> List[int]:
        """生成7个号码
        
        Args:
            context: 包含 position_frequency 的上下文
            
        Returns:
            7个号码
        """
        position_frequency = context.get('position_frequency', {})
        numbers = []
        
        for pos in range(1, 8):
            pos_freq = position_frequency.get(pos, {})
            
            if pos_freq:
                # 80% 概率选择高频号码
                if random.random() < 0.8:
                    top_numbers = sorted(pos_freq.keys(), key=lambda x: pos_freq[x], reverse=True)[:5]
                    numbers.append(random.choice(top_numbers))
                else:
                    numbers.append(random.choice(self.NUMBER_RANGE))
            else:
                numbers.append(random.choice(self.NUMBER_RANGE))
        
        return numbers
